### ISSUE TITLE
Doc retention fails inside trait bodies

### ISSUE THREAD

**Issue by jj101k** [glayzzle/php-parser#146]
On version 2.2.0, doc nodes are missing in trait bodies but are present in other comparable structures

Invocation:

```
var phpParser = require("php-parser")

var parser = new phpParser({
    parser: {
        debug: false,
        extractDoc: true,
    },
    ast: {
        withPositions: true,
    },
});
var tree = parser.parseCode(data, filename)
```

PHP source:

```
<?php
trait Foo {
    /**
     * @return bool
     */
    public function bar() {
        return true;
    }
}
```

If you change "trait" to "class", it works fine.

**Comment by jj101k** [comment]
Tree for trait:

```
{
  "kind": "program",
  "loc": {
    "source": null,
    "start": {
      "line": 1,
      "column": 0,
      "offset": 0
    },
    "end": {
      "line": 10,
      "column": 0,
      "offset": 111
    }
  },
  "children": [
    {
      "kind": "trait",
      "loc": {
        "source": null,
        "start": {
          "line": 2,
          "column": 0,
          "offset": 6
        },
        "end": {
          "line": 10,
          "column": 0,
          "offset": 111
        }
      },
      "name": "Foo",
      "extends": null,
      "implements": null,
      "body": [
        {
          "kind": "method",
          "loc": {
            "source": null,
            "start": {
              "line": 6,
              "column": 11,
              "offset": 65
            },
            "end": {
              "line": 8,
              "column": 5,
              "offset": 108
            }
          },
          "name": "bar",
          "arguments": [],
          "byref": false,
          "type": null,
          "nullable": false,
          "body": {
            "kind": "block",
            "loc": {
              "source": null,
              "start": {
                "line": 6,
                "column": 26,
                "offset": 80
              },
              "end": {
                "line": 8,
                "column": 5,
                "offset": 108
              }
            },
            "children": [
              {
                "kind": "return",
                "loc": {
                  "source": null,
                  "start": {
                    "line": 7,
                    "column": 8,
                    "offset": 90
                  },
                  "end": {
                    "line": 7,
                    "column": 20,
                    "offset": 102
                  }
                },
                "expr": {
                  "kind": "boolean",
                  "loc": {
                    "source": null,
                    "start": {
                      "line": 7,
                      "column": 15,
                      "offset": 97
                    },
                    "end": {
                      "line": 7,
                      "column": 19,
                      "offset": 101
                    }
                  },
                  "value": true
                }
              }
            ]
          },
          "isAbstract": false,
          "isFinal": false,
          "visibility": "public",
          "isStatic": false
        }
      ]
    }
  ],
  "errors": []
}
```

**Comment by jj101k** [comment]
Tree for class:

```
{
  "kind": "program",
  "loc": {
    "source": null,
    "start": {
      "line": 1,
      "column": 0,
      "offset": 0
    },
    "end": {
      "line": 10,
      "column": 0,
      "offset": 111
    }
  },
  "children": [
    {
      "kind": "class",
      "loc": {
        "source": null,
        "start": {
          "line": 2,
          "column": 0,
          "offset": 6
        },
        "end": {
          "line": 10,
          "column": 0,
          "offset": 111
        }
      },
      "name": "Foo",
      "isAnonymous": false,
      "extends": null,
      "implements": null,
      "body": [
        {
          "kind": "doc",
          "loc": {
            "source": null,
            "start": {
              "line": 3,
              "column": 4,
              "offset": 22
            },
            "end": {
              "line": 5,
              "column": 7,
              "offset": 53
            }
          },
          "isDoc": true,
          "lines": [
            "",
            "@return bool",
            ""
          ]
        },
        {
          "kind": "method",
          "loc": {
            "source": null,
            "start": {
              "line": 6,
              "column": 11,
              "offset": 65
            },
            "end": {
              "line": 8,
              "column": 5,
              "offset": 108
            }
          },
          "name": "bar",
          "arguments": [],
          "byref": false,
          "type": null,
          "nullable": false,
          "body": {
            "kind": "block",
            "loc": {
              "source": null,
              "start": {
                "line": 6,
                "column": 26,
                "offset": 80
              },
              "end": {
                "line": 8,
                "column": 5,
                "offset": 108
              }
            },
            "children": [
              {
                "kind": "return",
                "loc": {
                  "source": null,
                  "start": {
                    "line": 7,
                    "column": 8,
                    "offset": 90
                  },
                  "end": {
                    "line": 7,
                    "column": 20,
                    "offset": 102
                  }
                },
                "expr": {
                  "kind": "boolean",
                  "loc": {
                    "source": null,
                    "start": {
                      "line": 7,
                      "column": 15,
                      "offset": 97
                    },
                    "end": {
                      "line": 7,
                      "column": 19,
                      "offset": 101
                    }
                  },
                  "value": true
                }
              }
            ]
          },
          "isAbstract": false,
          "isFinal": false,
          "visibility": "public",
          "isStatic": false
        }
      ],
      "isAbstract": false,
      "isFinal": false
    }
  ],
  "errors": []
}
```

**Comment by jj101k** [comment]
Where the difference is:

Inside `read_trait()`, it does:

```
      body = this.next().read_class_body();
```

Inside `read_class()` it does:

```
    body = this.nextWithComments().read_class_body();
```

**Comment by ichiriac** [comment]
Hi @jj101k,

Thanks for the report. This issue is fixed as the document nodes were totally refactored. It will be released into the next version 3.0. Here the output of the trait (but works similar on any node) :

![image](https://user-images.githubusercontent.com/173203/38454260-e69e3950-3a63-11e8-8c61-5bd9a73fa141.png)

You can try the future release here : https://glayzzle.com/php-parser/
